design-toolkit/components/src/components/tree-view/index.tsx

```diff
@@ -0,0 +1,308 @@
+}
+
+function SelectionIcons({
+  selectionMode,
+  isSelected,
+}: { selectionMode: TreeSelectionMode; isSelected: boolean }) {
+  if (selectionMode === 'visibility') {
+    return <Icon>{isSelected ? <Show /> : <Hide />}</Icon>;
+  }
+
+  if (selectionMode === 'checkbox') {
+    return (
+      <Icon>{isSelected ? <CheckboxSelected /> : <CheckboxUnselected />}</Icon>
+    );
+  }
+
+  // selection === 'none'
+  return null;
+}
+
+function TreeNode<T extends TreeNodeType>({
+  children,
+  item,
+}: TreeNodeProps<T>) {
+  const { variant, selectionMode, allowsDragging, showRuleLines } =
+    useContext(TreeContext);
+
+  const {
+    getProps,
+    toggleSelect,
+    isSelected,
+    isFolder,
+    expand,
+    collapse,
+    isExpanded,
+  } = item;
+
+  const { onClick, ...props } = getProps();
+
+  const expander = isExpanded() ? (
+    <button type='button' onClick={expand} {...getProps()}>
+      <Icon>
+        <ChevronUp />
+      </Icon>
+    </button>
+  ) : (
+    <button type='button' onClick={collapse} {...getProps()}>
+      <Icon>
+        <ChevronDown />
+      </Icon>
+    </button>
+  );
+
+  const spacer = (
+    <span className='group-data-[variant=compact]:w-[15px] group-data-[variant=cozy]:w-[20px] group-data-[variant=tight]:w-[10px]' />
+  );
+
+  const folder = isFolder() ? expander : spacer;
+
+  const lines = Array.from({ length: item.getItemMeta().level }).map((_, i) => (
+    <span
+      key={i}
+      className={cn(
+        'relative',
+        'self-stretch group-data-[variant=compact]:w-[15px] group-data-[variant=cozy]:w-[20px] group-data-[variant=tight]:w-[10px]',
+        showRuleLines &&
+          (i !== item.getItemMeta().level - 1 ? 'vert-line' : 'tree-line'),
+      )}
+    />
+  ));
+
+  return (
+    <div
+      {...props}
+      data-variant={variant}
+      className={cn(
+        'group',
+        treeNodeStyles({
+          variant,
+          isDragTarget: item.isDragTarget() || item.isDraggingOver(),
+          isSelected: item.isSelected(),
+          isReadOnly: item.getItemData().isReadOnly,
+        }),
+      )}
+    >
+      {selectionMode !== 'none' && item.getItemData().isReadOnly ? (
+        <Icon>
+          <LockFill />
+        </Icon>
+      ) : (
+        <button
+          {...getProps()}
+          onClick={toggleSelect}
+          disabled={item.getItemData().isDisabled}
+        >
+          <SelectionIcons
+            selectionMode={selectionMode ?? 'none'}
+            isSelected={isSelected()}
+          />
+        </button>
+      )}
+      {lines}
+      {folder}
+      {item.getItemData().iconPrefix && (
+        <Icon>{item.getItemData().iconPrefix}</Icon>
+      )}
+      <div
+        className={cn(
+          'flex-1 items-center text-start',
+          'group-data-[variant=compact]:py-xs group-data-[variant=cozy]:py-s group-data-[variant=tight]:py-xxs',
+        )}
+      >
+        {item.getItemName()}
+        {item.getItemData().description && (
```

Comment:
Perf nit:

Can set `item.getItemData()` to a var and just re-use it. We may even be able to wrap it in a `useMemo`

```
const obj = useMemo(() => {
  return item.getItemData();
}, [item]);